libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxVoiceApi.kt

```diff
@@ -0,0 +1,112 @@
+     * @param callback VoiceCallback contains [VoiceState.VoiceFile]
+     */
+    override fun retrieveVoiceFile(voiceInstruction: VoiceInstructions, callback: VoiceCallback) {
+        val action = VoiceAction.PrepareVoiceRequest(voiceInstruction)
+        when (val result = VoiceProcessor.process(action)) {
+            is VoiceResult.VoiceRequest -> {
+                val mapboxSpeech = result.request
+                    .accessToken(accessToken)
+                    .language(language)
+                    .build()
+                mapboxSpeech.enqueueCall(object : Callback<ResponseBody> {
+                    override fun onResponse(
+                        call: Call<ResponseBody>,
+                        response: Response<ResponseBody>
+                    ) {
+                        val voiceAction = VoiceAction.ProcessVoiceResponse(response)
+                        when (val res = VoiceProcessor.process(voiceAction)) {
+                            is VoiceResult.Voice.Success -> {
+                                ThreadController.getMainScopeAndRootJob().scope.launch {
+                                    val voiceFile = generateVoiceFileFrom(res.data)
+                                    callback.onVoiceFileReady(VoiceState.VoiceFile(voiceFile))
+                                }
+                            }
+                            is VoiceResult.Voice.Failure -> {
+                                callback.onError(
+                                    VoiceState.VoiceError(res.error)
+                                )
+                            }
+                            is VoiceResult.Voice.Empty -> {
+                                callback.onError(
+                                    VoiceState.VoiceError("No data available")
+                                )
+                            }
+                        }
+                    }
+
+                    override fun onFailure(call: Call<ResponseBody>, t: Throwable) {
+                        callback.onError(
+                            VoiceState.VoiceError(t.localizedMessage ?: "Unknown")
+                        )
+                    }
+                })
+            }
+        }
+    }
+
+    /**
+     * Given the [Announcement] the method may cleanup any associated files previously generated.
+     * @param announcement
+     */
+    override fun clean(announcement: Announcement) {
+        announcement.file?.let {
+            delete(it)
+        }
+    }
+
+    private suspend fun generateVoiceFileFrom(data: ResponseBody): File =
+        withContext(ThreadController.IODispatcher) {
+            val cacheDirectory = context.applicationContext.cacheDir
+            val instructionsCacheDirectory = File(cacheDirectory, MAPBOX_INSTRUCTIONS_CACHE)
+                .also { it.mkdirs() }
+            File(instructionsCacheDirectory, "${retrieveUniqueId()}$MP3_EXTENSION")
```

Comment:
Is this file cleaned up anywhere? How large can it get?

Reply:
nevermind, after it is played it is deleted

Reply:
Refs. https://github.com/mapbox/mapbox-navigation-android/pull/3977/files#r574583524